### PR TITLE
Add algorithm to compute the Jacobian derivative of a link

### PR DESCRIPTION
**Pull request by diegoferigo**
This PR adds a new RBDA to compute the derivative of the free-floating Jacobian $\dot{J}_{W,L}$. If:

- $L$ is the frame of the link,
- $I$ is the frame of the input velocity representation used by the generalized velocity ${}^I \boldsymbol{\nu}$,
- $O$ is frame of the output velocity representation corresponding to the result of the product ${}^O \dot{J}_{W,L/I} \\; {}^I \boldsymbol{\nu}$,

we first compute the following full Jacobian derivative (with all columns filled, similarly to what was done in #121):

```math
{}^B \dot{J}_{W,\_/B} \in \mathbb{R}^{6\times(6+n)}
```

and then mask the joint-related columns using the support parent array $\kappa(i)$ of the i-th link associated to the frame $L$:

```math
{}^B \dot{J}_{W,L/B} = M(\kappa(i)) \; {}^B \dot{J}_{W,\_/B}
```

The mask in practice sets to zero all the joint-related columns corresponding to the chains not supporting the link $L$, i.e. not in the path $\pi_B(L)$.

All of this is computed with the new `jaxsim.rbda.jacobian.jacobian_derivative_full_doubly_left` function.

This is not enough, since we are interested to compute quantities like $\dot{J} \boldsymbol{\nu}$. Note that in this specific case, we already have this product computed by #127, it's only an use-case example. Though, we also need to change the output and input velocity representations to something else in order to support body-fixed, mixed, and inertial-fixed variants. Therefore, we need to properly convert ${}^B \dot{J}\_{W,L/B}$ to a generic ${}^O \dot{J}\_{W,L/I}$.

The most generic formulation of the change of representations of the free-floating Jacobian is the following[^1]:

```math
{}^O J_{W,L/I} = {}^O \mathbf{X}_B \; {}^B J_{W,L/B} \; \text{diag}({}^B \mathbf{X}_I,\: \mathbf{I}_n)
```

Therefore, if we want the derivative ${}^O \dot{J}\_{W,L/I}$, we need to also consider the derivatives of the different $\mathbf{X}$ matrices. The `jaxsim.api.link.jacobian_derivative` accounts for them. For this, we can use the following relation:

```math
{}^A \dot{\mathbf{X}}_B = {}^A \mathbf{X}_B \; {}^B \mathbf{v}_{A,B} \times
```

Finally:

```math
\begin{align}
{}^O \dot{J}_{W,L/I} =& {}^O \dot{\mathbf{X}}_B \; {}^B J_{W,L/B} \; \text{diag}({}^B \mathbf{X}_I,\: \mathbf{I}_n) \\
+& {}^O \mathbf{X}_B \; {}^B \dot{J}_{W,L/B} \; \text{diag}({}^B \mathbf{X}_I,\: \mathbf{I}_n) \\
+& {}^O \mathbf{X}_B \; {}^B J_{W,L/B} \; \text{diag}({}^B \dot{\mathbf{X}}_I,\: \mathbf{0}_n)
\end{align}
```

where $I \in \\{W, \\, B,\\, B[W]\\}$ and $O \in \\{W, \\, L,\\, L[W] \\}$.

[^1]: Diego Ferigo, Eq (2.28) pag. 67, _Simulation Architectures for Reinforcement Learning applied to Robotics_, Ph.D. thesis, [URL](https://github.com/diegoferigo/phd-thesis).

----
📚 Documentation preview 📚: https://jaxsim--169.org.readthedocs.build//169/

### CODE REVIEW

**Comment by diegoferigo** [comment]
cc @Giulero

File: tests/test_api_link.py [hunk 1]
```diff
@@ -211,3 +211,43 @@ def test_link_bias_acceleration(
...
+
+def test_link_jacobian_derivative(
+    jaxsim_models_types: js.model.JaxSimModel,
+    velocity_representation: VelRepr,
+    prng_key: jax.Array,
+):
+
+    model = jaxsim_models_types
+
+    key, subkey = jax.random.split(prng_key, num=2)
```
Comment (flferretti) [inline tests/test_api_link.py:224]: Just to be aligned with #166 
```suggestion
    _, subkey = jax.random.split(prng_key, num=2)
```

File: tests/test_api_link.py [hunk 2]
```diff
@@ -211,3 +211,43 @@ def test_link_bias_acceleration(
...
+        model=model,
+        key=subkey,
+        velocity_representation=velocity_representation,
+    )
+
+    # =====
+    # Tests
+    # =====
+
+    # Test only the last link of the model.
```
Comment (flferretti) [inline tests/test_api_link.py:235]: Why is that? Can we at least consider a subset?

File: src/jaxsim/api/link.py [hunk 3]
```diff
@@ -337,6 +338,187 @@ def velocity(
...
+        model=model,
+        joint_positions=data.joint_positions(),
+    )
+
+    # ... and extract the link Jacobian using the boolean support body array.
+    B_J_WL_B = jnp.hstack([jnp.ones(5), κb]) * B_J_full_WL_B
+
+    # Sum all the components that form the Jacobian derivative in the target
+    # input/output velocity representations.
+    O_J̇_WL_I = 0.0
```
Comment (flferretti) [inline src/jaxsim/api/link.py:514]: What about initializing with the final shape? So we can spot any shape error
```suggestion
    O_J̇_WL_I = jnp.zeros(shape=(6, 6 + model.dofs()))
```

**Review by flferretti — APPROVED** [review]
Great, thanks Diego! LGTM

**Comment by traversaro** [comment]
I am mostly travelling today and this PR is math heavy, so feel free to go ahead!

File: tests/test_api_link.py (hunk 2 above)
Comment (diegoferigo) [reply tests/test_api_link.py:235]: Yeah. I've just tried using vmap and the test still runs fast enough. I'll extend it to all links.

File: src/jaxsim/api/link.py (hunk 3 above)
Comment (diegoferigo) [reply src/jaxsim/api/link.py]: Done in cd8aa73

File: tests/test_api_link.py (hunk 2 above)
Comment (diegoferigo) [reply tests/test_api_link.py]: Done in 0bcae8b

**Comment by DanielePucci** [comment]
TL;DR Nice @diegoferigo! Just a curiosity: I imagine that we have a test checking the time derivative w.r.t. $J(t)$?

**Comment by diegoferigo** [comment]
> Just a curiosity: I imagine that we have a test checking the time derivative w.r.t. J(t)?

Since I implemented in #127 the computation of the product $(\dot{J} \boldsymbol{\nu}) \in \mathbb{R}^6$ in all the supported velocity representations, this PR includes a test that checks that such value is equal to the multiplication of the new $\dot{J} \in \mathbb{R}^{6\times(6+n)}$ with $\boldsymbol{\nu} \in \mathbb{R}^{6+n}$.

Out of curiosity, I took advantage of your question to check if we can compute $\dot{J}$ using AD instead of its analytical form. And it seems working fine :tada: I will extend the test in a new PR to add this check. It can be an interesting entrypoint for all those that want to exploit AD to compute gradients of quantities whose analytical form is too complicated to obtain.

**Comment by DanielePucci** [comment]
Super, thanks @diegoferigo !
